### PR TITLE
feat(forms): automatically support sticky forms

### LINKED ISSUE #11141: auto sticky form vars and auto prepare form vars + hook
It would make sense to let `elgg_view_form` trigger a hook to prepare form vars. They can be autopopulated by already existing sticky form vars for the given form. Maybe (if given an entity) data can also be preloaded based on the entity given... but that is not required at first

After getting form body the sticky form values can be cleared automatically.

In a action hook sticky form data could be saved based on the action/form name + cleared after a success response from the action file.

This makes sticky form usage very simple and prevents the requirement to execute prepare form functions like in blog..

### PR DESCRIPTION
**Pull request by jeabakker**
fixes #11141

ToDo:
- [x] add ignored fields support
- [x] global ignore more fields (`_route`, `_elgg_sticky_form_name`)
- [x] migrate registration form

### CODE REVIEW

File: docs/guides/events-list.rst [hunk 1]
```diff
@@ -911,6 +911,10 @@ Views
 **config, htmlawed** |results|
 	Filter the HTMLawed ``$config`` array.
 
+**form:prepare:fields, <form_name>** |results|
+	Prepare field values for use in the form. Eg. when editing a blog, fill this with the current values of the blog.
+	Sticky form values will automaticly be added to the field values (when available).
```
Comment (jdalsem) [inline docs/guides/events-list.rst:916]: automatically

File: engine/classes/Elgg/Router/Middleware/ActionMiddleware.php [hunk 2]
```diff
@@ -32,8 +42,41 @@ public function __invoke(\Elgg\Request $request) {
...
+	 * @return void
+	 * @since 5.0
+	 */
+	public function cleanupStickyValues(\Elgg\Event $event): void {
+		$response = $event->getValue();
+		if (!$response instanceof OkResponse) {
+			return;
+		}
+		
+		$form_name = $this->request->getParam('_elgg_sticky_form_name');
```
Comment (jdalsem) [inline engine/classes/Elgg/Router/Middleware/ActionMiddleware.php:79]: need to check if not empty... could be removed somewhere in between...

**Review by jdalsem — CHANGES_REQUESTED** [review]
also a sticky typo in your commit message :)

**Review by jdalsem — CHANGES_REQUESTED** [review]

File: views/default/resources/index.php [hunk 3]
```diff
@@ -5,7 +5,14 @@
 	$content = elgg_echo('index:content');
 } elseif (elgg_get_config('allow_registration')) {
 	$title = elgg_echo('register');
-	$content = elgg_view_form('register', ['ajax' => true]);
+	$content = elgg_view_form('register', [
+		'ajax' => true,
+		'sticky_enabled' => true,
```
[the remark this answers is not in the document]
Comment (jeabakker) [reply views/default/resources/index.php:10]: maybe not, but keeping all registration forms inline with each other

File: engine/classes/Elgg/Router/Middleware/ActionMiddleware.php [hunk 4]
```diff
@@ -32,8 +47,46 @@ public function __invoke(\Elgg\Request $request) {
...
+		$this->form_name = $this->request->getParam('_elgg_sticky_form_name');
+		
+		if (empty($this->form_name)) {
+			return;
+		}
+		
+		// add user and system ignored fields
+		$ignored_fields = (string) $this->request->getParam('_elgg_sticky_ignored_fields');
+		$ignored_fields = elgg_string_to_array($ignored_fields);
+		$ignored_fields = array_merge($ignored_fields, ['_elgg_sticky_form_name', '_elgg_sticky_ignored_fields', '_route']);
```
[the remark this answers is not in the document]
Comment (jeabakker) [reply engine/classes/Elgg/Router/Middleware/ActionMiddleware.php]: moved

**Review by jdalsem — APPROVED** [review]

File: engine/tests/phpunit/integration/Elgg/Forms/StickyFormsIntegrationTest.php [hunk 5]
```diff
@@ -41,6 +41,23 @@ protected function createService(Request $request): void {
 		]);
 	}
 	
+	public function testMakeStickyFormDefaultIgnoredFields() {
+		// request also container CSRF tokens, which shouldn't be made sticky
+		$request = $this->prepareHttpRequest('foo', 'POST', [
+			'name' => 'foo',
+			'_elgg_sticky_form_name' => 'foo', // added by sticky form support
```
[the remark this answers is not in the document]
Comment (jeabakker) [reply engine/tests/phpunit/integration/Elgg/Forms/StickyFormsIntegrationTest.php:48]: not in this test ;)

**Pull request merged**